<a href="https://colab.research.google.com/github/lucaslimacc/entrega_pandas_icd1/blob/master/pandas_intro_ex2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [1]:
!pip install pyarrow -q

In [2]:
import pandas as pd
import requests
import io

file_id = "19ydA23-eboyiKxN29JL1EQWW4CplTiKg"
url = f"https://drive.google.com/uc?export=download&id={file_id}"

response = requests.get(url)
df = pd.read_parquet(io.BytesIO(response.content))

## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [3]:
df_filt = df[df['Ano da Emenda'] == 2022]
df_filt.head(30)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
63473,202238950001,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,1,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950001
63474,202238950002,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,2,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,750141.00,750141.00,750141.00,0.00,0.0,0.00,202238950002
63475,202238950003,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,3,CRUZ DAS ALMAS - BA,2909802,CRUZ DAS ALMAS,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950003
63476,202238950004,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,4,IRAMAIA - BA,2914307,IRAMAIA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950004
63477,202238950005,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,5,MUCURI - BA,2922003,MUCURI,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950005
63478,202238950006,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,6,SERRA DO RAMALHO - BA,2930154,SERRA DO RAMALHO,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,535815.00,535815.00,535815.00,0.00,0.0,0.00,202238950006
63479,202238950007,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,7,SOUTO SOARES - BA,2930808,SOUTO SOARES,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950007
63480,202238950008,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,8,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,470837.00,470837.00,470837.00,0.00,0.0,0.00,202238950008
63481,202238950009,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,9,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,1100000.00,0.00,0.00,1100000.00,0.0,1100000.00,202238950009
63482,202238950010,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,10,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,8573048.00,8123048.00,8123048.00,450000.00,0.0,450000.00,202238950010


## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [5]:
# A sintaxe correta combina as condições dentro dos colchetes
df_norte = df[(df['Região'] == 'Norte') & (df['Valor Empenhado'] > 500000)]

# Mostra as 5 primeiras linhas e a contagem total
print(f"Quantidade de emendas: {len(df_norte)}")
df_norte.head(5)

Quantidade de emendas: 6230


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
18,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,MÂNCIO LIMA - AC,1200336,MÂNCIO LIMA,1200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,616610.0,0.0,0.0,0.0,0.0,616610.0,Sem informação
120,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,VITÓRIA DO JARI - AP,1600808,VITÓRIA DO JARI,1600000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,522384.4,0.0,0.0,0.0,0.0,522384.4,Sem informação
160,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTANA - AP,1600600,SANTANA,1600000,...,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,0,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,4443000.0,0.0,0.0,0.0,4443000.0,0.0,Sem informação
170,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,PALMEIRAS DO TOCANTINS - TO,1713809,PALMEIRAS DO TOCANTINS,1700000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,935350.0,0.0,0.0,0.0,0.0,935350.0,Sem informação
218,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,COSTA MARQUES - RO,1100080,COSTA MARQUES,1100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,549995.0,0.0,0.0,0.0,159996.0,389999.0,Sem informação


## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [6]:
df_uf = df[df['UF'].isin(['SÃO PAULO', 'RIO DE JANEIRO'])]

contagem = df_uf['UF'].value_counts()
print(contagem)

UF
SÃO PAULO         10435
RIO DE JANEIRO     6988
Name: count, dtype: int64


## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [8]:
print('Antes:')
print(df['Código da Emenda'].value_counts().head())

Antes:
Código da Emenda
Sem informação    17808
202341820007          2
202336610004          2
201833870003          2
201833870004          2
Name: count, dtype: int64


In [16]:
df['Código da Emenda'] = df['Código da Emenda'].replace('Sem informação', 'Não disponível')
print('Depois:')
print(df['Código da Emenda'].value_counts().head())

Depois:
Código da Emenda
Não disponível    17808
202341820007          2
202336610004          2
201833870003          2
201833870004          2
Name: count, dtype: int64


## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [26]:
df.loc[:, 'Nome Função'] = df['Nome Função'].str.upper()
df['Nome Função'].unique()

array(['URBANISMO', 'COMÉRCIO E SERVIÇOS', 'SAÚDE', 'EDUCAÇÃO',
       'AGRICULTURA', 'ASSISTÊNCIA SOCIAL', 'DESPORTO E LAZER',
       'DEFESA NACIONAL', 'SANEAMENTO', 'DIREITOS DA CIDADANIA',
       'GESTÃO AMBIENTAL', 'CULTURA', 'CIÊNCIA E TECNOLOGIA',
       'SEGURANÇA PÚBLICA', 'ORGANIZAÇÃO AGRÁRIA', 'COMUNICAÇÕES',
       'ADMINISTRAÇÃO', 'HABITAÇÃO', 'INDÚSTRIA', 'TRABALHO', 'MÚLTIPLO',
       'TRANSPORTE', 'PREVIDÊNCIA SOCIAL', 'ENCARGOS ESPECIAIS',
       'RELAÇÕES EXTERIORES', 'ENERGIA', 'ESSENCIAL À JUSTIÇA'],
      dtype=object)

## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [20]:
print('Total de duplicatas:', df.duplicated().sum())
print('Total de linhas antes:', len(df))

Total de duplicatas: 63
Total de linhas antes: 88659


In [22]:
df_sem_dup = df.drop_duplicates()
print('Total de duplicatas:', df_sem_dup.duplicated().sum())
print('Total de linhas depois:', len(df_sem_dup))

Total de duplicatas: 0
Total de linhas depois: 88596


## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro?

In [39]:
print('Total de duplicatas:' , df.duplicated(subset = ['Código da Emenda', 'Município']).sum())
print('Total de linhas antes:' , len(df))

Total de duplicatas: 12915
Total de linhas antes: 88659


In [40]:
df_sem_dup2 = df.drop_duplicates(subset = ['Código da Emenda', 'Município'])

In [41]:
print('Total de linhas depois: ') , len(df_sem_dup2)

Total de linhas depois: 


(None, 75744)

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [27]:
print(df.isnull().sum())

Código da Emenda                          0
Ano da Emenda                             0
Tipo de Emenda                            0
Código do Autor da Emenda             15961
Nome do Autor da Emenda               15961
Número da emenda                          0
Localidade de aplicação do recurso        0
Código Município IBGE                     0
Município                                 0
Código UF IBGE                            0
UF                                        0
Região                                    0
Código Função                             0
Nome Função                               0
Código Subfunção                          0
Nome Subfunção                            0
Código Programa                           0
Nome Programa                             0
Código Ação                               0
Nome Ação                                 0
Código Plano Orçamentário                 0
Nome Plano Orçamentário                   0
Valor Empenhado                 

# Código do Autor da Emenda             15961
# Nome do Autor da Emenda               15961

essas são as linhas que possuem o maior número de nulos

## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [29]:
df['Código do Autor da Emenda'] = df['Código do Autor da Emenda'].fillna(-1)
print(df.isnull().sum())

Código da Emenda                          0
Ano da Emenda                             0
Tipo de Emenda                            0
Código do Autor da Emenda                 0
Nome do Autor da Emenda               15961
Número da emenda                          0
Localidade de aplicação do recurso        0
Código Município IBGE                     0
Município                                 0
Código UF IBGE                            0
UF                                        0
Região                                    0
Código Função                             0
Nome Função                               0
Código Subfunção                          0
Nome Subfunção                            0
Código Programa                           0
Nome Programa                             0
Código Ação                               0
Nome Ação                                 0
Código Plano Orçamentário                 0
Nome Plano Orçamentário                   0
Valor Empenhado                 

In [34]:
df['Nome do Autor da Emenda'] = df['Nome do Autor da Emenda'].fillna('desconhecido')
print(df.isnull().sum())

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             0
Código UF IBGE                        0
UF                                    0
Região                                0
Código Função                         0
Nome Função                           0
Código Subfunção                      0
Nome Subfunção                        0
Código Programa                       0
Nome Programa                         0
Código Ação                           0
Nome Ação                             0
Código Plano Orçamentário             0
Nome Plano Orçamentário               0
Valor Empenhado                       0
Valor Liquidado                       0
Valor Pago                            0


## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [42]:
df["Ano da Emenda"].dtype

dtype('int64')

In [47]:
df['Ano da Emenda'] = df['Ano da Emenda'].astype(str)
print('Tipo de Emenda:', df["Ano da Emenda"].dtype)

Tipo de Emenda: object


## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [49]:
df["Valor Empenhado"].dtype

dtype('float64')

# A coluna Valor Empenhado já está em float

## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [53]:
df.groupby('Região')['Valor Pago'].sum()

,Valor Pago
Região,
Centro-Oeste,1.037094e+10
Exterior,2.572317e+07
Múltiplo,3.010563e+10
Nacional,7.886148e+09
Nordeste,4.644963e+10
Norte,2.035103e+10
Sudeste,3.427210e+10
Sul,1.648893e+10


In [55]:
media_regiao = df.groupby('Região')['Valor Pago'].sum()
print(media_regiao.sort_values(ascending=False))

Região
Nordeste        4.644963e+10
Sudeste         3.427210e+10
Múltiplo        3.010563e+10
Norte           2.035103e+10
Sul             1.648893e+10
Centro-Oeste    1.037094e+10
Nacional        7.886148e+09
Exterior        2.572317e+07
Name: Valor Pago, dtype: float64


## 13) Calcule a **quantidade** de emendas por **Nome Função**. Mostre as 5 funções com mais emendas.

In [63]:
mais_emendas = df["Nome Função"].value_counts()
print(mais_emendas.sort_values(ascending=False).head(5))
print(f"Função: {mais_emendas.idxmax()}")
print(f"Quantidade: {mais_emendas.max()}")

Nome Função
SAÚDE                 37068
URBANISMO              8663
EDUCAÇÃO               8108
AGRICULTURA            4940
ENCARGOS ESPECIAIS     4610
Name: count, dtype: int64
Função: SAÚDE
Quantidade: 37068


## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [70]:
df["Saldo a Pagar"] = df["Valor Empenhado"] - df["Valor Pago"]

df[["Valor Empenhado", "Valor Pago", "Saldo a Pagar"]].head(5)

,Valor Empenhado,Valor Pago,Saldo a Pagar
0,700000.00,0.0,700000.00
1,487500.00,0.0,487500.00
2,100000.00,0.0,100000.00
3,243750.00,0.0,243750.00
4,1658333.94,0.0,1658333.94


## 15) Elimine a(s) coluna(s) do dataframe que eestá totalmente vazia.

In [66]:
df.dropna(axis=1, how="all", inplace=True)

In [67]:
print(df.shape)

(88659, 29)


In [73]:
df.isnull()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1,Saldo a Pagar
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
88654,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
88655,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
88656,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
88657,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.

In [68]:
df.to_json("emendas.json", orient="records", force_ascii=False, indent=2)